File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head==None or head.next==None:
            return None
        t1=head
        c=0
        while t1!=None:
            t1=t1.next
            c=c+1
        rem=c-n
        t1=head
        t2=head
        while(rem>0):
            t2=t1
            t1=t1.next
            rem-=1
        
        if t1.next==None:
            t2.next=None
        else:
            t1.val=t1.next.val
            t1.next=t1.next.next
        return head

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len=0;
        ListNode* t=head;
        while(t!=NULL){
            len++;
            t=t->next;
        }
        int actnodemoves=len-n,prevnode=actnodemoves-1;
        if(prevnode<0){
            return head->next;
        }
        ListNode *t1=head,*t2=head;
        while(actnodemoves--){
            t1=t1->next;
        }
        while(prevnode--){
            t2=t2->next;
        }
        t2->next=t1->next;
        return head;
    }
};